From: Chengming Zhou <chengming.zhou@linux.dev>
Subject: mm-zswap-use-only-one-pool-in-zswap-v3
Date: Tue, 25 Jun 2024 12:39:05 +0800

fix error handling again in zswap_pool_create(), per Yosry

Link: https://lkml.kernel.org/r/20240625-zsmalloc-lock-mm-everything-v3-2-ad941699cb61@linux.dev
Signed-off-by: Chengming Zhou <chengming.zhou@linux.dev>
Reviewed-by: Nhat Pham <nphamcs@gmail.com>
Acked-by: Yosry Ahmed <yosryahmed@google.com>
Cc: Chengming Zhou <zhouchengming@bytedance.com>
Cc: Johannes Weiner <hannes@cmpxchg.org>
Cc: Minchan Kim <minchan@kernel.org>
Cc: Sergey Senozhatsky <senozhatsky@chromium.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/zswap.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

--- a/mm/zswap.c~mm-zswap-use-only-one-pool-in-zswap-v3
+++ a/mm/zswap.c
@@ -268,7 +268,7 @@ static struct zswap_pool *zswap_pool_cre
 	pool->zpool = zpool_create_pool(type, name, gfp);
 	if (!pool->zpool) {
 		pr_err("%s zpool not available\n", type);
-		return NULL;
+		goto error;
 	}
 	pr_debug("using %s zpool\n", zpool_get_type(pool->zpool));
 
@@ -303,7 +303,8 @@ ref_fail:
 error:
 	if (pool->acomp_ctx)
 		free_percpu(pool->acomp_ctx);
-	zpool_destroy_pool(pool->zpool);
+	if (pool->zpool)
+		zpool_destroy_pool(pool->zpool);
 	kfree(pool);
 	return NULL;
 }
_
